Added ReactFiberStack shared by ReactFiberContext and ReactFiberHostContext

bvaughn opened the pull request:
#### Overview

`ReactFiberStack` is the new underlying stack construct used by `ReactFiberContext` and `ReactFiberHostContext`. The goal is to simplify the 2 context managers _and_ to add earlier and better dev-warnings when we pop unexpectedly.

This change set causes a lot of tests to fail as we are currently popping too many times and/or in the wrong order. I'm pushing this branch now to share with Sebastian as he is working on a related cleanup in `beginWork`.

#### Work needing to be done for this branch
- [x] Track Fibers with unique host contexts to avoid over-popping; (see [this](https://github.com/facebook/react/pull/8611#issuecomment-268596186))
- [x] Tighten up the `Fiber | null` type in `ReactFiberStack` by making sure that we always pass a `Fiber` down; currently this is not the case for `pushHostContainer` and `popHostContainer`.
- [x] Verify that no new test failures are caused by this refactor.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberStack.js:
```diff
@@ -0,0 +1,102 @@
+    valueStack[index] = emptyObject;
+
+    if (__DEV__) {
+      fiberStack[index] = emptyObject;
```
Why emptyObject instead of null on both of these? We could use ES2015 `.fill()` here if available.

Another thing we could do if we're worried about leakage is to store the max index of the previous run and then only clear the delta between the previous two runs. That way if we end up refilling the stack every time we rerender we don't have to fill it twice.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberStack.js:
```diff
@@ -0,0 +1,102 @@
+
+import type { Fiber } from 'ReactFiber';
+
+export type StackCursor = {|
```
We should make this generic `type StackCursor<T> = {| current: T |}`.

You will have to make some unsound things when you pop here but then everywhere else is fine.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberStack.js:
```diff
@@ -0,0 +1,102 @@
+
+let index = -1;
+
+exports.createCursor = function() : StackCursor {
```
If you add an initial default value, Flow will let you make this generic.

`function<T>(defaultValue : T) : StackCursor<T> {`

When you can invoke it with `createCursor((null : ?string))` to get a typed version of the stack cursor.

bvaughn replied:
Passing a default value feels a bit awkward since (at least for now) none of the cursors have meaningful defaults. But the generics benefit is nice! So sure! 😄

sebmarkbage commented:
In ReactFiberScheduler we call unwindContext followed up unwindHostContext. The point of these is to reset the context of both to where the error boundary is. However, this will break with your change because the popping needs to be interleaved not sequential.

I.e. you need to pop *both* (or all three I guess) until we reach the desired error boundary. I think we need to have an `unwindWork` that does only the normal popping that would happen in completeWork based on the tag. This is effectively what goes into a `finally` block in a programming language.

bvaughn commented:
Regarding `unwindContext` and `unwindHostContext`, it seems like the simplest solution would be to rename `unwindHostContext` to something like `unwindContexts` and add a case for `ClassComponent` that pop context providers. Not sure what the benefit of it being a separate module is at the moment since the scheduler is the only thing that unwinds. What's the motivation there?

bvaughn commented:
Another TODO item for me: In `pushHostContainer`, [we don't push if the current host context and the next/child host context are equal](https://github.com/facebook/react/blob/afd4a54d2a1b95a3e6df3052c4af63bb4154f05b/src/renderers/shared/fiber/ReactFiberHostContext.js#L113-L115). In order to avoid over-popping, I need to add another cursor to track the fibers we push for. I have a meeting shortly but I'll do that once I'm out.

sebmarkbage commented:
For some reason I thought that unwindHostContext wasn't in the scheduler. Ignore the comment about separate module.

sebmarkbage commented:
I was thinking that maybe we should tag the host components with a different tag if they don't need to change context. That way we know when we complete that we don't need to pop it. Similarly, the next time we update, we can pick the fast path of skipping checking for context changes.

bvaughn commented:
Where would this tagging be done? Seems inappropriate for `ReactFiberHostContext` to modify a `Fiber` in any way. I was planning on tracking which host `Fiber`s provided unique context with another `StackCursor`. In theory this should work but my first attempt causes 3 Jest tests to hang. Still investigating.

If you'd like to chat about ideas here ping me on Messenger. 😄

sebmarkbage commented:
The thing calling `pushHostContext` can mutate the tag from `beginWork`. Indeterminate and Coroutine components already do this. I'm not sure the ReactFiberContext and ReactFiberHostContext modules need to exist as abstractions. They can also just be inlined at their callsites. Don't let the abstraction get in your way.

gaearon commented on src/renderers/shared/fiber/ReactFiberHostContext.js:
```diff
@@ -35,129 +43,87 @@ module.exports = function<T, P, I, TI, C, CX>(
 
   function getRootHostContainer() : C {
-    if (rootInstance == null) {
+    if (rootInstanceStackCursor.current == null) {
```
Minor nit: could you read `.current` once? I don't know if compiler is smart enough to know it hasn't changed.

bvaughn replied:
Absolutely.

gaearon commented:
The issue I'm noticing is that adding

```js
      static childContextTypes = {
        foo: React.PropTypes.number
      };
      getChildContext() {
        return {foo: 42};
      }
```

to some of the "broken" components in `ReactErrorBoundaries-test` breaks some tests with unexpected pops. It's not just `BrokenRender` that fails this way but `BrokenComponentWillMount`, `BrokenComponentWillReceiveProps` and a few more.

It's likely the same issue as #8604 but worth noting it doesn't only happen for `render` method.

bvaughn commented:
There's a kind of obvious design flaw with this single-stack-with-shared-cursors approach that I managed to overlook until a half hour ago.
```
Action      Cursor A    Cursor b    Stack
------      --------    --------    -----
(init)      null        null        []
push(a,1)   1           null        [1]
push(b,2)   1           2           [1,2]
pop(b)      1           1           [1]     <- b is invalid
pop(a)      null        1           []      <- b is still invalid
```
 Irrespective of mismatched pushes/pops, this breaks a lot of things. I'm reevaluating the best way to proceed at the moment. I think Sebastian was a big fan of a single stack but I'm not currently able to visualize a way for it to work at the moment.

sebmarkbage commented:
The thing that is suppose to go on the stack is the previous value of the cursor. Not the value you push.

```
Action      Cursor A    Cursor b    Stack
------      --------    --------    -----
(init)      null        null        []
push(a,1)   1           null        [null]
push(b,2)   1           2           [null,null]
pop(b)      1           null        [null]     <- b is valid
pop(a)      null        null        []      <- a and b is valid
```

bvaughn commented:
Ahhh, that makes sense. Thanks @sebmarkbage.

sebmarkbage commented:
The mental model is:

```js
var current = null;
function work(foo) {
  // Begin Work
  var prevOnTheStack = current;
  current = getContext(foo);
  try {
    processChildren(foo);
    // Complete Work
    completeSomething();
  } finally {
    // Complete Work and Unwind Context 
    current = prevOnTheStack;
  }
}

bvaughn commented:
As of my most recent push, all tests that were previously passing are now once again passing on this branch.

sebmarkbage commented:
CI is failing for some reason. Try running `grunt build` locally?

bvaughn commented:
Yeah, I noticed that. Running locally I see a lot of garbage in the console like
> Error message "Missing owner for string ref %s" cannot be found. The current React version and the error map are probably out of sync. Please run `gulp react:extract-errors` before building React.

...but then it ultimately finishes with
> Done, without errors.

Nuked `node_modules` and did a fresh `yarn install` in case a dependency had changed. Still works for me. I'm going to tell Circle to rebuild without cache.

bvaughn commented:
I am able to reproduce the error locally by manually running the gulp task `./node_modules/.bin/gulp react:extract-errors`. After looking at the node failing, I've determined that it's caused by my use of an exact object type for `StackCursor`:

```js
// This works
export type StackCursor<T> = { current: T };

// This causes the error
export type StackCursor<T> = {| current: T |};
```

bvaughn commented:
CI is getting further now but it's ultimately failing for a different reason. Running the same task locally (`./scripts/circleci/test_coverage.sh`) fails against both my branch and master.

bvaughn commented:
Had an informative chat with Ben about this. The failing Grunt task was caused by the fact that I've added new invariants/warnings in this PR and updated my local error codes JSON. This in turn caused our `ReactPropTypesProduction` test to fail but in a non-obvious way. Rather than showing the error message from `ReactPropTypesProduction`, _another error_- `TypeError: process.cwd is not a function`- was shown. This is because we're mocking out the `process` object that Jest is trying to use. Commenting out the call to `process.cwd` in Jest revealed the underlying error code message.

This further confused me for a while because Grunt failed for me locally on master as well. That turned out to be caused by a caching issue which was fixed with PR #8622.

In general we have a test flakiness problem that will bite us when we next update the error codes (before the next release). That flakiness issue will need to be addressed in a subsequent PR. I don't mind taking that task on once these context PRs have been wrapped up. 😄 

For now, CI passes again for this branch. Yay.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberContext.js:
```diff
@@ -24,20 +25,22 @@ var {
-const contextStack : Array<Object> = [];
-const didPerformWorkStack : Array<boolean> = [];
+let contextStackCursor : StackCursor<?Object> = createCursor((null: ?Object));
+let didPerformWorkStackCursor : StackCursor<?boolean> = createCursor((null: ?boolean));
```
If you set the default here to `(false : boolean)` you can get rid of the nullable type.

bvaughn replied:
I thought about doing that, but then I'd lose the ability to differentiate between "empty" and default-state. Which isn't really important in this case but... slightly more compelling is that it would also mean the cursor would have asymmetric behavior- `false` by default, and then `null` after the stack gets reset (since we don't track and fall back to the default value). It seemed worth doing the cast to avoid that, but I don't feel strongly about it.

sebmarkbage replied:
It won't have asymmetric behavior if you fix the reset below. :)

bvaughn replied:
You are correct of course! 😁  That turnery was a holdover from when I was pushing/popping incorrectly. D'oh.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberStack.js:
```diff
@@ -0,0 +1,101 @@
+};
+
+exports.reset = function<T>(
+  cursor : StackCursor<T>,
```
Passing a cursor here seems like an odd API since it's invalid to only reset one cursor. You have to reset all at once. It also causes the unsound assignment below.

Currently this only works because index is already at `-1` at the second call to reset so the loop is unnecessary.

I'd recommend just dropping the `cursor` argument and reset them separately if you think it's necessary to reset them.

bvaughn replied:
Sure. It's a little weird either way I guess. I'll drop the cursor from `reset` and just manually clear the `current` pointer :)

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberStack.js:
```diff
@@ -0,0 +1,101 @@
+
+exports.push = function<T>(
+  cursor : StackCursor<T>,
+  value : any,
```
This can be `value : T` for a bit stronger guarantees that ensures that we don't push unexpected values.

bvaughn replied:
Good catch.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberStack.js:
```diff
@@ -0,0 +1,101 @@
+
+  cursor.current = index > 0
+    ? valueStack[index]
+    : (null : any);
```
This is unsound since it doesn't reset to the defaultValue. However, I don't think this should ever be possible since the first value that goes into the value stack __is__ the defaultValue. So you shouldn't be able to pop further than that.

sebmarkbage replied:
This should just be:

```js
cursor.current = valueStack[index];
```

Otherwise you're never reading index 0.

bvaughn replied:
Good suggestion.

bvaughn commented on src/renderers/shared/fiber/ReactFiberContext.js:
```diff
@@ -62,7 +66,7 @@ exports.getMaskedContext = function(workInProgress : Fiber) {
 
 exports.hasContextChanged = function() : boolean {
-  return index > -1 && didPerformWorkStack[index];
+  return Boolean(didPerformWorkStackCursor.current);
```
Ok. :) I'll go ahead and make that change then. I don't feel strongly about it.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberContext.js:
```diff
@@ -129,13 +131,15 @@ exports.pushContextProvider = function(workInProgress : Fiber, didPerformWork :
 
 exports.resetContext = function() : void {
-  index = -1;
+  reset();
```
We now have two calls to reset in a row because the only thing that calls this is:

https://github.com/facebook/react/blob/master/src/renderers/shared/fiber/ReactFiberScheduler.js#L157-L160

So we could just call `reset();` at the top of that function instead.

bvaughn replied:
Seems slightly weird to bypass the contexts and talk directly to the stack they use, no?

But sure. I can change that.

bvaughn commented:
Thanks for the review Sebastian!!
